### PR TITLE
Support zipTree(Map) to supply encoding

### LINKED ISSUE #11983: German umlaute are replaced with cryptic characters after unpacking zip-archive
I'm trying to extract a zip-file, which includes files with a german umlaut in its filename.
After unpacking the zip-file, all umlaute are replaced with cryptic characters.

**My Gradle Task**

```groovy
task extractZip(dependsOn: clean){
    doLast{
        copy {
            FileTree updateZip = zipTree(file("$projectDir/UmlauteInside.zip"))
            from updateZip
            into file("$buildDir/extractedZipfile")
        }
    }
}
```
| Filenames inside the Zip-Archive | Filenames after extracting the Zip-Archive|
| ------------------------------------ | ------------------------------------ |  
|  Umlaute-ÄÖÜ.txt | Umlaute-Ž™š.txt|  
|  Ä.txt | Ž.txt |  
|  Ö.txt | ™.txt |  
|  Ü.txt | š.txt |  

### Expected Behavior
The umlaute should be taken over as they are

### Current Behavior
The umlaute are replaced with special or cryptic characters

### Steps to Reproduce 
Example Project on GitHub: https://github.com/leozent/gradleExtractZip
Usage on commandline: `.\gradlew extractZip`

### Your Environment
Build scan URL: https://scans.gradle.com/s/zxfl4sltrmzqi

### PR DESCRIPTION
**Pull request by jeffalder**
Fixes #11983 

### Context
Users can specify the charset for _output_ zip files, it makes sense to allow them to specify the charset for _input_ zip files.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#git-commit---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective **NOTE** There are no integTest setups for these changes
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [x] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [x] Ensure that tests pass locally: `./gradlew <changed-subproject>:check`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by jeffalder** [comment]
I'm completely confused by the CI results. Several failed for "Connection reset" which seems unrelated, and there was also the "binary compatibility" issue, which I'm not 100% clear how to address. Since I didn't remove any existing API methods, this _should_ be compatible. I would think. I don't know how to get the diff report, or info on any resolutions.

**Comment by jeffalder** [comment]
The "binary compatibility" issue, when run against my local, generates a pile of this message:
```
2020-04-05T16:43:11.937-0700 [DEBUG] [japicmp.model.JavaObjectSerializationCompatibility] Failed to get serialVersionUid from class org.gradle.api.file.Rel
        at japicmp.model.JavaObjectSerializationCompatibility.computeSerialVersionUid(JavaObjectSerializationCompatibility.java:49)
        at japicmp.model.JavaObjectSerializationCompatibility.extractSerialVersionUid(JavaObjectSerializationCompatibility.java:24)
        at japicmp.model.JApiClass.<init>(JApiClass.java:79)
```
Which I also don't understand.

**Comment by jeffalder** [comment]
@big-guy Thanks for the tags; any tips on getting past CI here?

**Comment by jeffalder** [comment]
I fixed up the`@since` and `@Incubating` that broke the binary compatibility issue. The other targets that failed seem to be network-related?

**Comment by jeffalder** [comment]
My CI is still failing for remote builds that I can't do anything about :-(

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/delegates/ProjectDelegate.kt [hunk 1]
```diff
@@ -154,6 +154,9 @@ abstract class ProjectDelegate : Project {
     override fun zipTree(zipPath: Any): FileTree =
         delegate.zipTree(zipPath)
 
+    override fun zipTree(args: MutableMap<String, *>): FileTree =
```
Comment (vlsi) [inline subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/delegates/ProjectDelegate.kt:157]:
```suggestion
    override fun zipTree(args: Map<String, *>): FileTree =
```

Why is it mutable?

File: subprojects/core/src/main/java/org/gradle/api/internal/file/DefaultFileOperations.java [hunk 2]
```diff
@@ -153,7 +155,33 @@ public ConfigurableFileTree fileTree(Map<String, ?> args) {
...
+    }
+
+    @Override
+    public FileTree zipTree(Map<String, ?> args) {
+        Object zipPath = args.get("file");
+        if (zipPath == null) {
+            throw new InvalidUserDataException("zipTree requires the 'file' argument.");
+        }
+
+        Object metadataCharset = args.containsKey("metadataCharset") ? args.get("metadataCharset") : null;
```
Comment (vlsi) [inline subprojects/core/src/main/java/org/gradle/api/internal/file/DefaultFileOperations.java:168]:
```suggestion
        Object metadataCharset = args.get("metadataCharset");
```

Is `containsKey` required?

Comment (jeffalder) [reply subprojects/core/src/main/java/org/gradle/api/internal/file/DefaultFileOperations.java:168]: It's not required, thanks!

File: subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/delegates/ProjectDelegate.kt (hunk 1 above)
Comment (jeffalder) [reply subprojects/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/support/delegates/ProjectDelegate.kt:157]: I think I used IDEA's auto-creation that set this up. I don't see a reason to support MutableMaps.

**Comment by jeffalder** [comment]
force-pushed to get sign-off.

**Comment by jeffalder** [comment]
@vlsi - I adopted your changes.

**Comment by vlsi** [comment]
@jeffalder , thanks. 💪 

The change you make looks good to me, however, I'm not sure if it is in line or not with the way Gradle public API is designed.
I just don't know if `Map` for two-argument method is ok or not. I see you might anticipate "more arguments added in the future", however, I just don't know. 🤷‍♂️  

One more item is I don't like `resources.findResource("cp437filenames.zip")` because there are often issues like "resource not found due to wrong working directory" (e.g. when the test is executed from IDE), and it is not clear how to create such a zip file (e.g. if someone wants to add new encoding later).
If I were you, I would create cp437 zip in the test method (then it would be easy to understand what makes the archive different, and it would probably be easier to maintain).

Then I know Gradle team appreciates **end-to-end** tests. You claim that your PR improves `zipTree(...)`, however, you add no `build.gradle`-like test that would ensure your modification is **usable** from Groovy (and/or Kotlin) DSL :(
Do you think you could add `build.gradle` or (`build.gradle.kts` like) test?

PS. I'm not on a Gradle team, and their opinion might easily differ.

**Comment by jeffalder** [comment]
@vlsi - Thanks for the thoughtful comments!

*zipTree uses `Map` for multiple arguments*: I was mirroring the signatures of `fileTree`, which has `ConfigurableFileTree fileTree(Map<String, ?> args);` and `ConfigurableFileTree fileTree(Object baseDir);`. I know that many provide an `Action` as a second argument; I considered that, but this is usually nested inside something else (e.g., `from(zipTree(...`), so formatting would get kind of messy:
```
  from(zipTree("cp437filenames.zip") {
    metadataCharset = "cp437"
  })
```

I can be convinced either way, but I hope that someone from the gradle team will weigh in.

*hardcoded zip name*: There are two other calls to that same function in the test file, so this test would not be the only one to fail. I opted to include the actual file for several reasons:
1. The file is small (530 bytes); larger files would give me pause.
2. The file is a direct repro case from a github issue. I tagged the test with `@Issue` so people should be able to find it.
3. The zip file is not easily recreated from the command-line on most systems.

*end-to-end*: There are no tests in `core/integTest` that verify the results of `zipTree` directly; most are testing the zip or copy tasks. I did, however, manage to add an integTest that uses the Ant 'zip' task. I also added verification that the encoding of the file name matches expected values.

This also pointed out that I was missing the `Script` overload, so that was good.

Ultimately I'm a little uncomfortable relying on the fact that gradle can unzip its own zip files. That seems like something could be missed and be ignored.

**Comment by jeffalder** [comment]
I'm closing this PR. Feel free to re-engage if it becomes useful or interesting.

**Pull request closed without merging**